docs/source/worker-state.rst

```diff
@@ -0,0 +1,433 @@
+   nowhere to be found in memory anymore) and reschedule task x somewhere else on the
+   cluster. Task x will transition to ``cancelled(flight)`` on worker A.
+7. If the scheduler randomly chooses worker A to compute task X, the task will
+   transition to ``resumed(flight->waiting)``.
+8. When, *and only when*, the TCP socket from A to B collapses (e.g. due to timeout),
+   the task will transition to ``waiting`` and will be eventually recomputed on A.
+
+.. important::
+
+   You always have *at most* one :meth:`~WorkerBase.compute` or
+   :meth:`~WorkerBase.gather_dep` asyncio task running for any one given key; you
+   never have both.
+
+
+Task state mapping between Scheduler and Worker
+-----------------------------------------------
+
+The task states on the scheduler and the worker are different, and their mapping is
+somewhat nuanced:
+
++------------------+-----------------------+-------------------------+
+| Scheduler states | Typical worker states | Edge case worker states |
++==================+=======================+=========================+
+| - released       | - (unknown)           | - released              |
+| - waiting        |                       | - cancelled             |
+| - no-worker      |                       |                         |
++------------------+-----------------------+-------------------------+
+| - processing     | - waiting             | - resumed(waiting)      |
+|                  | - ready               |                         |
+|                  | - constrained         |                         |
+|                  | - executing           |                         |
+|                  | - long-running        |                         |
++------------------+-----------------------+-------------------------+
+| - memory         | - memory              | - error                 |
+|                  | - fetch               | - missing               |
+|                  | - flight              | - resumed(fetch)        |
++------------------+-----------------------+-------------------------+
+| - erred          | - error               |                         |
++------------------+-----------------------+-------------------------+
+
+In addition to the above states, a worker may not know about a specific task at all.
+The opposite, where the worker knows about a task but it is nowhere to be found on the
+scheduler, happens exclusively in the case of :ref:`cancelled-tasks`.
+
+There are also *race conditions* to be considered, where a worker (or some workers) know
+something before the scheduler does, or the other way around. For example,
+
+- A task will always transition from ``executing`` to ``memory`` on the worker before
+  it can transition from ``processing`` to ``memory`` on the scheduler
+- A task will always transition to ``released`` or ``forgotten`` on the scheduler first,
+  and only when the message reaches the worker it will be released there too.
+
+
+Flow control
```

Comment:
Reading this section of the docs makes me so happy with the design of the worker state machine. Having state transformation strongly separated out from IO and concurrency like this is so nice, and such a big improvement. Nice work!